### PR TITLE
G35: Bed tramming assistant

### PR DESCRIPTION
**Pull request by rmcbc**
### Requirements

Add the G35 command to read bed corners to help adjust screws.

### Description

The objective is to measure all bed screws with BLTouch or other device for auto bed leveling and suggest the number of turns that each screw must have to be leveled with the others.
This is important to avoid Z movements to correct the bad leveling of the bed.

### Benefits

Avoid Z movements with a bad leveled bed and providing a easy way to help leveling by adjusting the screws.

### Example

Recv: Adjust Front Left
Recv:     Direction = Clockwise
Recv:    Full turns = 0
Recv:       Minutes = 30
Recv: Adjust Rear Left
Recv:     Direction = Clockwise
Recv:    Full turns = 0
Recv:       Minutes = 11
Recv: Adjust Rear Right
Recv:     Direction = Clockwise
Recv:    Full turns = 0
Recv:       Minutes = 14

### CODE REVIEW

**Comment by rmcbc** [comment]
Please test it and say what you think about this new feature.

**Comment by vivian-ng** [comment]
@rmcbc Just to clarify, will this work with manual probe? Or only with BLTouch and inductive/capacitive probes? If there are certain probes it does not work with, it may be good to add in a sanity check.

**Comment by rmcbc** [comment]
> @rmcbc Just to clarify, will this work with manual probe? Or only with BLTouch and inductive/capacitive probes? If there are certain probes it does not work with, it may be good to add in a sanity check.

@vivian-ng this should work with all kind of probes, but of course the better the probe the best result you get.

**Comment by randellhodges** [comment]
I find this an interesting addition. I am going to give it a try later and see how it works.

Will it also work for 3 point bed leveling?

```
  //               Default value is 0 (Clockwise M3), most printers use an
  //               M3 screw and turning the knob clockwise decrease distance.
```

Do you mean decrease the distance between the bed and the nozzle? I think for typical bed slingers, like the Ender 3, counter (anti) clockwise allows the bed to move up, which decreases the distance.

It might seem minor, but you might say `decrease the distance between nozzle and bed`.

Also, if this really takes off, you might look into separating `SCREWS_TILT_ADJUST_THREAD` into a direction and thread pitch setting. Then if someone had a odd ball adjustment screw, that person could enter whatever. Or do a combination. Allow a selection like you have, with the ability to specify each.

First things first, let's see how well it works as you have it :)

**Comment by rmcbc** [comment]
> I find this an interesting addition. I am going to give it a try later and see how it works.
> 
> Will it also work for 3 point bed leveling?
> 

Yes, it should, because it compares the difference from the first point to the others.

> ```
>   //               Default value is 0 (Clockwise M3), most printers use an
>   //               M3 screw and turning the knob clockwise decrease distance.
> ```
> 
> Do you mean decrease the distance between the bed and the nozzle? I think for typical bed slingers, like the Ender 3, counter (anti) clockwise allows the bed to move up, which decreases the distance.
> 
> It might seem minor, but you might say `decrease the distance between nozzle and bed`.
> 

No, the distance between the bed and the bed support.

> Also, if this really takes off, you might look into separating `SCREWS_TILT_ADJUST_THREAD` into a direction and thread pitch setting. Then if someone had a odd ball adjustment screw, that person could enter whatever. Or do a combination. Allow a selection like you have, with the ability to specify each.
> 
> First things first, let's see how well it works as you have it :)

Thanks and hope to see the results.

**Comment by rmcbc** [comment]
> @rmcbc Just to clarify, will this work with manual probe? Or only with BLTouch and inductive/capacitive probes? If there are certain probes it does not work with, it may be good to add in a sanity check.

@vivian-ng I'm sorry you are right I should verify if there is a probe in the system, because it's a must have. Thanks.

**Comment by randellhodges** [comment]
> > I find this an interesting addition. I am going to give it a try later and see how it works.
> > Will it also work for 3 point bed leveling?
> 
> Yes, it should, because it compares the difference from the first point to the others.
> 
> > ```
> >   //               Default value is 0 (Clockwise M3), most printers use an
> >   //               M3 screw and turning the knob clockwise decrease distance.
> > ```
> > 
> > 
> > Do you mean decrease the distance between the bed and the nozzle? I think for typical bed slingers, like the Ender 3, counter (anti) clockwise allows the bed to move up, which decreases the distance.
> > It might seem minor, but you might say `decrease the distance between nozzle and bed`.
> 
> No, the distance between the bed and the bed support.

I see. Maybe `decrease the bed height`? Sorry if I seem fixated on this, I can see others getting confused, especially with the international community. A small wording change will make it a bit less ambiguous. 

It could also just be me, since I think of it as tramming and not bed leveling, thus in my head, the reference point is the nozzle :)
> 
> > Also, if this really takes off, you might look into separating `SCREWS_TILT_ADJUST_THREAD` into a direction and thread pitch setting. Then if someone had a odd ball adjustment screw, that person could enter whatever. Or do a combination. Allow a selection like you have, with the ability to specify each.
> > First things first, let's see how well it works as you have it :)
> 
> Thanks and hope to see the results.

**Comment by rmcbc** [comment]
> I see. Maybe `decrease the bed height`? Sorry if I seem fixated on this, I can see others getting confused, especially with the international community. A small wording change will make it a bit less ambiguous.

Done, thanks.

**Comment by randellhodges** [comment]
I played around with it a bit today and found how much I hate 4 point bed leveling. My comments here are only applicable to beds with 4 leveling screws.

I have an Ender 3. My leveling screws are 33 mm inset from the corners.

Based on your earlier comment, I changed my `SCREWS_TILT_ADJUST_THREAD` to 2, but it kept recommending a clockwise change, which would tighten the screw, thus causing that position to go lower. Looking at a G30 probe at that same point, I actually need to losen the screw to allow the spring to push the bed "up", so I actually needed a `SCREWS_TILT_ADJUST_THREAD` of 3.

The inherit problem with a 4 point bed leveling system is that, adjustment at one corner will impact measurements at all the other corners. Sorry that I don't have hard numbers to provide, but I would run it and get a recommendation of 15 minutes (so I assume 1/4 a turn). I would give it a 1/4 turn in the correct direction, run again and it would recommend a 17 minute turn, in the same direction.

I started turning only one screw, then G35 again (home, probe all 4, etc) and it seemed that I never got any closer to 0.

I did start off with a very level bed.

I feel there is real potential here. Maybe someone else will have better luck or can recommend some specific optimizations. There might be some stuff in the 4 z stepper code of G34 that might help overcome the issues with leveling a plane on 4 corners.

**Comment by rmcbc** [comment]
@randellhodges First of all thanks for testing it.

> The inherit problem with a 4 point bed leveling system is that, adjustment at one corner will impact measurements at all the other corners. Sorry that I don't have hard numbers to provide, but I would run it and get a recommendation of 15 minutes (so I assume 1/4 a turn). I would give it a 1/4 turn in the correct direction, run again and it would recommend a 17 minute turn, in the same direction.
> 
> I started turning only one screw, then G35 again (home, probe all 4, etc) and it seemed that I never got any closer to 0.
> 
> I did start off with a very level bed.
> 
> I feel there is real potential here. Maybe someone else will have better luck or can recommend some specific optimizations. There might be some stuff in the 4 z stepper code of G34 that might help overcome the issues with leveling a plane on 4 corners.

I understand your problem, I have a bed with a glass on it and it's almost impossible to get all corners perfectly leveled, but for me this G35 it's a great help because it guides me, in a 4 screws bed, which screws to tight or lose to get it more or less leveled.
Normally I stop when I get values around 15 minutes (that means 1/4 turn).
This way the bed is good and makes AUTO BED LEVELING almost no correction moves on Z to compensate.
I don't think it's possible to calibrate to 0 all screws, but this helps you to get closer to it.

**Comment by AnHardt** [comment]
A plane is completely defined by 3 points (https://en.wikipedia.org/wiki/Coplanarity). Defining more points over constrains the system. Then, in general, only bending the surface helps to meet all defined points.
It's like the impossibility to build a not wobbling, four-legged chair. Without bending something by loading the seat it will always wabble a tiny bit.

4 point hardware bed leveling is a hardware error.

**Comment by rmcbc** [comment]
> A plane is completely defined by 3 points (https://en.wikipedia.org/wiki/Coplanarity). Defining more points over constrains the system. Then, in general, only bending the surface helps to meet all defined points.
> It's like the impossibility to build a not wobbling, four-legged chair. Without bending something by loading the seat it will always wabble a tiny bit.
> 
> 4 point hardware bed leveling is a hardware error.

Hi @AnHardt , if I understand correctly, actually the G35 command used on a 4 screws bed only ask you to adjust 3 of them, because one is the reference for the others.
See this as a helper to know the direction and how much you need to turn the screws to have a bed almost leveled. In my experience it´s almost impossible to have a perfectly leveled bed. But this can help.

**Comment by randellhodges** [comment]
> A plane is completely defined by 3 points (https://en.wikipedia.org/wiki/Coplanarity). Defining more points over constrains the system. Then, in general, only bending the surface helps to meet all defined points.
> It's like the impossibility to build a not wobbling, four-legged chair. Without bending something by loading the seat it will always wabble a tiny bit.
> 
> 4 point hardware bed leveling is a hardware error.

I would agree, but the majority of 3d printer manufacturers appear not to agree. They put 4 screws in that need to be adjusted and that was the basis for my comment.

I might change my hardware over to a 3 point leveling setup, once I find one that is robust. The one I tried was a bit thin on the metal and seemed to flex more than I cared for. A lot of people, probably the most that would benefit from this PR, are probably going to have 4 screws to turn. (My guess anyway).

**Comment by thinkyhead** [comment]
Since none of my printers have a probe that can reach as close to the right side of the bed as they can to the left side of the bed, the measurements can't be taken where the screws are. Will this compensate for the fact that my probe is reading the bed height a few cm to the left of the right screws?

**Comment by thinkyhead** [comment]
> 4 point hardware bed leveling is a hardware error.

Well, we're tilting a flat plane by putting the 4 corners at slightly different heights, and as long as there is some play at the 4 points where the screws mate with the bed you won't get any real distortion of the bed from these adjustments. But if you have a very rigid connection you could warp the aluminium and glass by some measurable amount.

**Comment by rmcbc** [comment]
> Since none of my printers have a probe that can reach as close to the right side of the bed as they can to the left side of the bed, the measurements can't be taken where the screws are. Will this compensate for the fact that my probe is reading the bed height a few cm to the left of the right screws?

@thinkyhead this system is more accurate if the probe can reach the screws, but it also helps if you make the probe at the same distance from the screw.

Example in a 220 x 220 bed:
- Front Left 30,30; 
- Front Right 190,30; 
- Rear Right 190,190; 
- Rear Left 30,190.

This way all points are at same distance from the screws, is not as accurate as exactly on top, but it also helps.

**Comment by rmcbc** [comment]
@thinkyhead do you need anything else to include this PR in bugfix-2.0.x?

**Comment by rvermootenct** [comment]
I've physically tested this and it works nicely.
I also read through the documentation and believe it reads well.
 @thinkyhead have you been busy or is there something else you're looking for?

**Comment by InsanityAutomation** [comment]
A few notes from my point of view. In most cases, you will be instructing users to place this directly over the bed leveling knobs under the bed. The comment should reflect that a bit better for Cartesian XZ machines. For CartesianXY/CoreXY it should spell out as close to straight lines between rods as possible.

I see it does handle 3 positions, which is good as a few actually have that!

For Screw thread, id recommend prompting them to directly input the pitch. You can call out common thread pitches in the comment, however there is a strong likelihood someone will have a thread pitch not included here at some point. For tight memory boards, this half of the output could be optional and only included if a pitch is defined.

See G34 for provisions around tool change
```cpp
      tool_change(old_tool_index, (true
        #if ENABLED(PARKING_EXTRUDER)
          && false // Fetch the previous toolhead
        #endif
      ));
```

As probe offsets can change at runtime, making positions unreachable, Gcode syntax should be added to redefine positions (See M422). It seems to me that G34 and G35 share essentially the same set of values and can share the same set gcode and eeprom storage, you're simply enabling one or the other based on how many Z steppers you have.

**Comment by rmcbc** [comment]
@InsanityAutomation thanks for your notes, they are very interesting.
I'm going to focus the changes in the Screw thread part because seems to me that is a great improvement, the other part related to probing position I've to analyze it further.

**Comment by InsanityAutomation** [comment]
No problem, let me know if you get stuck. I can either give you some pointers or send a PR towards you're branch.

**Comment by hussainsail2002** [comment]
can this feature be used along with the level corners feature ?. The level corners feature we can adjust the number of leveling points between 3 and 4. Also we can set the inset from the edges.

**Comment by InsanityAutomation** [comment]
@hussainsail2002 this requires a probe and gives precise measurements and turn amounts, but on the flip side may have trouble fitting on a melzi. So they're best kept separate and this aligns with G34 better as it needs to use the same type of configurable probe points.

**Comment by qwewer0** [comment]
I get the following errors after adding and modifying the needed files:
```
identifier \"set_axis_is_not_at_home\" is undefined
expression must have a constant value -- cannot call non-constexpr function "__builtin_trunc" (declared implicitly)
```

Marlin 2.0.5.3, SKR Mini E3 v1.2

**Comment by NickdeK** [comment]
Getting the following error:
```
Marlin\src\gcode\bedlevel\G35.cpp: In static member function 'static void GcodeSuite::G35()':
Marlin\src\gcode\bedlevel\G35.cpp:214:7: error: 'set_axis_is_not_at_home' was not declared in this scope
       set_axis_is_not_at_home(Z_AXIS);
```

**Comment by qwewer0** [comment]
Error messages:

```
G35.cpp
expression must have a constant value -- cannot call non-constexpr function "__builtin_trunc"(declared implicitly) [44, 17]
'set_axis_is_not_at_home' was not declared in this scope [214, 7]
identifier "set_axis_is_not_at_home" is undefined [214, 7]
```

SKR Mini E3 v1.2, latest Bugfix-2.0.x

**Comment by rmcbc** [comment]
@qwewer0 error in compilation fixed. Needed to merge some changes in bugfix-2.0.x

**Comment by qwewer0** [comment]
> @qwewer0 error in compilation fixed. Needed to merge some changes in bugfix-2.0.x

Works great, thanks.

**Comment by thinkyhead** [comment]
I've got a few tricky issues to debug this weekend, but I will review this soon and get it in shape for final testing and merge.

**Comment by rmcbc** [comment]
@thinkyhead thanks for your help, it's very important to put some of your magic to make it perfect.

**Comment by chestwood96** [comment]
Just tried this out and it is pretty neat. (I was using G29 J and the bed visualizer to do something similar before but this is a lot more streamlined)

I second @randellhodges suggestion of adding a separate pitch and direction setting.

Also would be cool if you could add distance, degrees and fraction of a turn to the output.

**Comment by NickdeK** [comment]
I've tried to test this with my **Ender 3 Pro** + **SKR Mini E3** + **BLTouch** but it's not working. G35 starts with homing which seems to work fine, goes to the bottom left corner, probes it and throws the following in the terminal
```
Recv: Probing failed: Point (Front Right) 1 X=205.00 Y=30.00
Recv: G35 aborted.
Recv: echo:busy: processing
[...]
Recv: echo:busy: processing
[...]
Communication timeout while idle, trying to trigger response from printer. Configure long running commands or increase communication timeout if that happens regularly on specific commands or long moves.
```
After that it looks like it's homing again and that's it.

The probing point of the corner looks right to me, it probes and stops but it results in a probing failed error.

**Comment by randellhodges** [comment]
@NickdeK That usually means the difference between the center of the bed and the corner is too much. Try raising the corner up and try again.

**Comment by NickdeK** [comment]
> @NickdeK That usually means the difference between the center of the bed and the corner is too much. Try raising the corner up and try again.

```
Send: G29 T
Recv: 
Recv: Bed Topography Report:
Recv: 
Recv:     (  5,220)                                                              (230,220)
Recv:         0       1       2       3       4       5       6       7       8       9
Recv:  9 | +0.115  +0.085  +0.055  +0.043  +0.025  +0.010  +0.027  +0.007  +0.007  +0.007
Recv:    |
Recv:  8 | +0.092  +0.065  +0.040  +0.025  -0.005  +0.020  +0.027  +0.023  +0.023  +0.023
Recv:    |
Recv:  7 | +0.077  +0.053  +0.035  +0.035  -0.008  +0.013  +0.017  +0.015  +0.015  +0.015
Recv:    |
Recv:  6 | +0.102  +0.070  +0.055  +0.035  -0.005  +0.010  +0.013  -0.005  -0.005  -0.005
Recv:    |
Recv:  5 | +0.097  +0.065  +0.045  +0.035   0.000  +0.013  +0.010  -0.010  -0.010  -0.010
Recv:    |
Recv:  4 | +0.102  +0.060  +0.033  +0.023  -0.005  +0.007  +0.010  -0.023  -0.023  -0.023
Recv:    |
Recv:  3 | +0.112  +0.082  +0.050  +0.047  +0.005  +0.020  +0.020  +0.005  +0.005  +0.005
Recv:    |
Recv:  2 | +0.112  +0.080  +0.055  +0.053  +0.025  +0.040  +0.043  +0.023  +0.023  +0.023
Recv:    |
Recv:  1 | +0.135  +0.085  +0.062  +0.057  +0.027  +0.040  +0.033  +0.027  +0.027  +0.027
Recv:    |
Recv:  0 | +0.162  +0.127  +0.107  +0.102  +0.072  +0.085  +0.082  +0.085  +0.088  +0.090
Recv:         0       1       2       3       4       5       6       7       8       9
````
Is +0.162 to much of a difference to compensate?

Also, the probe does trigger in the bottom left corner and the z movement stops, after that it throws the error.

**Comment by randellhodges** [comment]
Not sure if your reply was meant to be snarky or not, but it came across that way. I am going to chalk it up to the internet and assume you did not mean to be.

If you want to see if +0.162 is too much to compensate, execute a G30 over bed screw. If it is still fails, then it is a problem with something other than G35 code.

I have seen the same probe fail error with G34 and with G30 but only when a corner was way off, hence the reason for my suggestion.

I'm not sure G30 disables bed leveling before it does it's thing. I'd disable bed leveling, G28, then do a G30 X33 Y33 (example for the bottom left Ender 3 screw location) and see what it says.

Do note, if you do it like this and make an adjustment, you probably want to G28 again before measuring again.

**Comment by NickdeK** [comment]
> Not sure if your reply was meant to be snarky or not, but it came across that way. I am going to chalk it up to the internet and assume you did not mean to be.
> 
> If you want to see if +0.162 is too much to compensate, execute a G30 over bed screw. If it is still fails, then it is a problem with something other than G35 code.
> 
> I have seen the same probe fail error with G34 and with G30 but only when a corner was way off, hence the reason for my suggestion.
> 
> I'm not sure G30 disables bed leveling before it does it's thing. I'd disable bed leveling, G28, then do a G30 X33 Y33 (example for the bottom left Ender 3 screw location) and see what it says.
> 
> Do note, if you do it like this and make an adjustment, you probably want to G28 again before measuring again.

Oh no, I'm absolutely not trying to be snarky! I really like this feature and wanted to test it and help where possible. Apologies for the used words, I can see now how that could have sounded snarky.

I will try and rule out hardware issues with your suggestions later today. Thanks for fast reply :)

**Comment by NickdeK** [comment]
So I found the problem. The problem is not at all related to the error printed in the terminal, the error is coming from a wrong set coordinate for the second probe spot. On the Ender 3 the BLTouch is usually installed on the left side (as with mine) and so you have to take into account that the probe coordinate for the right side should take into account the X offset. I had this set to 210, but the probe can't reach that spot. Maybe the error messaging can be improved here?

The following issue now occurs; the hotend travels to the second point after it probed the first point. ~~However, it doesn't move the Z up enough before it travels to the second point.~~ This results in the probe not being able to deploy fully and the machine going into a halted state. ~~To be clear, it does move up but not enough for it to fully deploy the probe.~~ On second glance, when the hotend travels to the next point the Z moves down during the travel. I'm not entirely sure, but it looks like it's around the ammount of the Z offset noted below 🤔 

In the code I see you use the `Z_CLEARANCE_BETWEEN_PROBES` variable, I have set this to 2 which works when running G29 with UBL enabled. I have a Z offset of -2.3, I'm going to check if increasing the `Z_CLEARANCE_BETWEEN_PROBES` to 5 will resolve this issue.

Setting the `Z_CLEARANCE_BETWEEN_PROBES` to 5 resolves the issue. I think something is not right here, a `Z_CLEARANCE` of 2 should be enough with my current leveled bed. Although I can understand that if you use G35 your bed might be off a lot more and 2 might not be enough, should G35 have a seperate `Z_CLEARANCE` variable?

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -762,14 +762,14 @@
...
   /**
    * Screw thread:
-   *   M3: 0 = Clockwise, 1 = Counter-Clockwise
-   *   M4: 2 = Clockwise, 3 = Counter-Clockwise
-   *   M5: 4 = Clockwise, 5 = Counter-Clockwise
+   *   M3: 30 = Clockwise, 31 = Counter-Clockwise
+   *   M4: 40 = Clockwise, 41 = Counter-Clockwise
+   *   M5: 50 = Clockwise, 51 = Counter-Clockwise
    *
    * Default is Clockwise M3 (0). Most printers use M3
```
Comment (randellhodges) [inline Marlin/Configuration_adv.h]: Should now read: `Default is Clockwise M3 (30)` or maybe just omit the value?

File: Marlin/Configuration_adv.h [hunk 2]
```diff
@@ -762,14 +762,14 @@
...
    * screws, with clockwise turns moving the bed down.
```
Comment (randellhodges) [inline Marlin/Configuration_adv.h]: Does `bed down` mean away from the nozzle? Thinking about printers where the bed moves up/down vs the nozzle moving up/down. Am I overthinking it?

Comment (AnHardt) [reply Marlin/Configuration_adv.h]: Only if you'd consider the nozzle hanging tip top below the bed, `bed down` would decrease the distance of bed and nozzle. If the bed or the nozzle is the motorized part does not matter.

**Comment by ETE-Design** [comment]
@rmcbc Would be great if you could make a menu for it for LCD Screen with GFX Graphics :-D

**Comment by thisiskeithb** [comment]
This doesn't compile after recent changes. See PR https://github.com/MarlinFirmware/Marlin/issues/18382:

```prolog
Marlin/src/gcode/bedlevel/G35.cpp: In static member function 'static void GcodeSuite::G35()':
Marlin/src/gcode/bedlevel/G35.cpp:187:3: error: jump to label 'EXIT_G35'
  187 |   EXIT_G35:
      |   ^~~~~~~~
Marlin/src/gcode/bedlevel/G35.cpp:88:10: note:   from here
   88 |     goto EXIT_G35;
      |          ^~~~~~~~
Marlin/src/gcode/bedlevel/G35.cpp:117:8: note:   crosses initialization of 'bool err_break'
  117 |   bool err_break = false;
      |        ^~~~~~~~~
In file included from Marlin/src/gcode/bedlevel/../../inc/../core/boards.h:24,
                 from Marlin/src/gcode/bedlevel/../../inc/MarlinConfigPre.h:35,
                 from Marlin/src/gcode/bedlevel/../../inc/MarlinConfig.h:28,
                 from Marlin/src/gcode/bedlevel/G35.cpp:23:
Marlin/src/gcode/bedlevel/G35.cpp:96:50: note:   crosses initialization of 'const bool leveling_was_active'
   96 |     TERN_(RESTORE_LEVELING_AFTER_G35, const bool leveling_was_active = planner.leveling_active);
      |                                                  ^~~~~~~~~~~~~~~~~~~
Marlin/src/gcode/bedlevel/../../inc/../core/macros.h:400:26: note: in definition of macro 'THIRD'
  400 | #define THIRD(a,b,c,...) c
      |                          ^
Marlin/src/gcode/bedlevel/../../inc/../core/macros.h:195:29: note: in expansion of macro '___TERN'
  195 | #define __TERN(T,V...)      ___TERN(_CAT(_NO,T),V)  // Prepend '_NO' to get '_NOT_0' or '_NOT_1'
      |                             ^~~~~~~
Marlin/src/gcode/bedlevel/../../inc/../core/macros.h:194:29: note: in expansion of macro '__TERN'
  194 | #define _TERN(E,V...)       __TERN(_CAT(T_,E),V)    // Prepend 'T_' to get 'T_0' or 'T_1'
      |                             ^~~~~~
Marlin/src/gcode/bedlevel/../../inc/../core/macros.h:193:29: note: in expansion of macro '_TERN'
  193 | #define TERN_(O,A)          _TERN(_ENA_1(O),,A)     // OPTION converted to A or '<nul>'
      |                             ^~~~~
Marlin/src/gcode/bedlevel/G35.cpp:96:5: note: in expansion of macro 'TERN_'
   96 |     TERN_(RESTORE_LEVELING_AFTER_G35, const bool leveling_was_active = planner.leveling_active);
      |     ^~~~~
Compiling .pio/build/LPC1768/src/src/gcode/config/M200-M205.cpp.o
Compiling .pio/build/LPC1768/src/src/gcode/config/M217.cpp.o
*** [.pio/build/LPC1768/src/src/gcode/bedlevel/G35.cpp.o] Error 1
```

**Comment by KyleLampkin** [comment]
Hello all,
It would be helpful too if you added a menu item under the advanced settings where you could run this with out needing a terminal so that anyone could benefit from it similar to the way the M48 repeatability test is done.

**Comment by thisiskeithb** [comment]
> Hello all,
> It would be helpful too if you added a menu item under the advanced settings where you could run this with out needing a terminal so that anyone could benefit from it similar to the way the M48 repeatability test is done.

Please put in a [Feature Request](https://github.com/MarlinFirmware/Marlin/issues/new?assignees=&labels=&template=feature_request.md&title=%5BFR%5D+%28feature+request+title%29).

**Comment by KyleLampkin** [comment]
I will do, I thought that ETE-Design had already done that. My apologies. I think it's a great tool to help level the bed to the same plane as the nozzle is on.Great work by the way.

**Comment by SysGh-st** [comment]
I've been curious about this feature (G35). So far what I've tested: Hooks up Pronterface to the printer and issues the G35 command. Printer then performs all the probing (BLTouch) and returns to the middle. I had to set everything 45mm inwards, as the bltouch sits so far away to the left. So far all probe points work with no errors. 

But... 
I get no feedback in Pronterface, no fails or any instructions on how much to turn each screw. It just remains silent. 
I understand that the display/UI features isn't implemented yet, but.. it should echo back to the serial console, right? Did I miss something?

EDIT 2020-08-06:
I did figure it out that it needs DEBUG_LEVELING_FEATURE enabled after reading through the code.
Suggestion: add to the comment in configuration_adv.h that this feature currently needs DEBUG_LEVELING_FEATURE to be useful in the current state. Should save people some headache.

**Comment by ETE-Design** [comment]
@rmcbc Any GFX or something, so it is possible to do it All from LCD?

**Comment by thisiskeithb** [comment]
@ETE-Design: There is an open feature request to have `G35` work from the LCD controller: https://github.com/MarlinFirmware/Marlin/issues/18489

**Comment by qwewer0** [comment]
<s>Is there a way to make G35 to probe 5 points, with the first probing point to be at the middle of the bed, and the other four at the corners?

I tried to do it by adding a fifth point, and shift them so the middle one is the first, but it probed the middle twice then act as the middle is the first corner then it would try to move to the next "corner" relative to the middle, so it hits the max software endstops. (I stopped the printer at that point)
```
  #define TRAMMING_POINT_XY { { 115, 115 }, { 30, 200 }, { 30,  30 }, { 200, 30 }, { 200, 200 } }

  #define TRAMMING_POINT_NAME_1 "Middle"
  #define TRAMMING_POINT_NAME_2 "Back-Left"
  #define TRAMMING_POINT_NAME_3 "Front-Left"
  #define TRAMMING_POINT_NAME_4 "Front-Right"
  #define TRAMMING_POINT_NAME_5 "Back-Right"
```
</s>


Seems to be working now, even if I don't know why it did not before...

**Comment by NeoMod** [comment]
I decided to test this feature after fine-tuning my printer, and I must say that it works really well. Thank you for this feature!

It took 4/5 measurements (using a BLTouch clone) to obtain a " 0 turns " indication on all points; very easy to use and intuitive in the configuration.

**Comment by thinkyhead** [comment]
Does anyone still consider it important to set the Z axis un-trusted at the end? The PR #22060 proposes to remove this, which seems sensible since other operations at the end of `G35` seem to assume you can proceed to normal operation. For example, it re-enables bed leveling, which makes little sense if the user is then going to be prompted to re-home Z. What do you think?

**Pull request merged**